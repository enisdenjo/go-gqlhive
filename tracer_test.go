package gqlhive

import (
	"context"
	"os"
	"testing"

	"github.com/99designs/gqlgen/client"
	"github.com/99designs/gqlgen/graphql/handler"
	"github.com/domonda/go-types/nullable"
	"github.com/enisdenjo/go-gqlhive/fixtures/todos/graph"
	"github.com/gkampitakis/go-snaps/snaps"
)

func TestMain(t *testing.M) {
	v := t.Run()
	snaps.Clean(t)
	os.Exit(v)
}

func TestCreatedReports(t *testing.T) {
	var queries = []string{
		"{ todos { id } }",
		`{
			todos {
				id
				user {
					name
					id
				}
				text
				done
			}
		}`,
		`mutation CreateTodo {
			createTodo(input: { text: "Check Mail", userId: "u0" }) {
				id
				text
				user {
					name
				}
				done
			}
		}`,
		`mutation CreateTodo {
			createTodo(input: { userId: "u0", text: "Check Mail" }) {
				id
			}
		}`,
		`{
			todos(sortBy: NAME_DESC, condition: { searchText: "test" }) {
				id
			}
		}`,
		`{
			todos(condition: { statuses: [DONE, ASSIGNED] }) {
				id
			}
		}`,
		`{
			todos(condition: { userStatus: AVAILABLE }) {
				id
			}
		}`,
		`{
			todos(condition: { user: { name: "deep" } }) {
				id
			}
		}`,
		`{
			todos {
				...TodoFragment
			}
		}
		fragment TodoFragment on Todo {
			id
			text
			user {
				... on User {
					id
					name
				}
			}
			done
		}
		`,
		`query Todos($searchText: String) {
			todos(condition: { searchText: $searchText }) {
				id
			}
		}`,
	}

	for _, query := range queries {
		t.Run(query, func(t *testing.T) {
			server := handler.NewDefaultServer(graph.NewExecutableSchema(graph.Config{Resolvers: &graph.Resolver{}}))

			var sentReport *Report
			server.Use(NewTracer(
				"<token>",
				WithGenerateID(func(operation string, operationName nullable.TrimmedString) string {
					return "id"
				}),
				WithSendReport(func(ctx context.Context, endpoint, token string, report *Report) error {
					for _, info := range report.OperationInfos {
						info.Timestamp = -1
						info.Execution.Duration = -1
					}
					sentReport = report
					return nil
				}),
			))

			res := map[string]any{}
			client.New(server).MustPost(query, &res)

			snaps.MatchJSON(t, sentReport)
		})
	}
}
