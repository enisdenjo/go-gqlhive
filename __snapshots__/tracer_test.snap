
[TestCreatedReports/{_todos_{_id_}_} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Todo.id"
   ],
   "operation": "{ todos { id } }"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/{____todos_{_____id_____user_{______name______id_____}_____text_____done____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Todo.id",
    "Todo.user",
    "User.name",
    "User.id",
    "Todo.text",
    "Todo.done"
   ],
   "operation": "{\n\t\t\ttodos {\n\t\t\t\tid\n\t\t\t\tuser {\n\t\t\t\t\tname\n\t\t\t\t\tid\n\t\t\t\t}\n\t\t\t\ttext\n\t\t\t\tdone\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/mutation_CreateTodo_{____createTodo(input:_{_text:_"Check_Mail",_userId:_"u0"_})_{_____id_____text_____user_{______name_____}_____done____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Mutation.createTodo",
    "Mutation.createTodo.input",
    "NewTodo.text",
    "String",
    "NewTodo.userId",
    "ID",
    "Todo.id",
    "Todo.text",
    "Todo.user",
    "User.name",
    "Todo.done"
   ],
   "operation": "mutation CreateTodo {\n\t\t\tcreateTodo(input: { text: \"Check Mail\", userId: \"u0\" }) {\n\t\t\t\tid\n\t\t\t\ttext\n\t\t\t\tuser {\n\t\t\t\t\tname\n\t\t\t\t}\n\t\t\t\tdone\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/mutation_CreateTodo_{____createTodo(input:_{_userId:_"u0",_text:_"Check_Mail"_})_{_____id____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Mutation.createTodo",
    "Mutation.createTodo.input",
    "NewTodo.userId",
    "ID",
    "NewTodo.text",
    "String",
    "Todo.id"
   ],
   "operation": "mutation CreateTodo {\n\t\t\tcreateTodo(input: { userId: \"u0\", text: \"Check Mail\" }) {\n\t\t\t\tid\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/{____todos(sortBy:_NAME_DESC,_condition:_{_searchText:_"test"_})_{_____id____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Query.todos.sortBy",
    "TodosSortBy.NAME_DESC",
    "Query.todos.condition",
    "TodosCondition.searchText",
    "String",
    "Todo.id"
   ],
   "operation": "{\n\t\t\ttodos(sortBy: NAME_DESC, condition: { searchText: \"test\" }) {\n\t\t\t\tid\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/{____todos(condition:_{_statuses:_[DONE,_ASSIGNED]_})_{_____id____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Query.todos.condition",
    "TodosCondition.statuses",
    "TodosConditionStatus.DONE",
    "TodosConditionStatus.ASSIGNED",
    "Todo.id"
   ],
   "operation": "{\n\t\t\ttodos(condition: { statuses: [DONE, ASSIGNED] }) {\n\t\t\t\tid\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/{____todos(condition:_{_userStatus:_AVAILABLE_})_{_____id____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Query.todos.condition",
    "TodosCondition.userStatus",
    "TodosConditionUserStatus.AVAILABLE",
    "Todo.id"
   ],
   "operation": "{\n\t\t\ttodos(condition: { userStatus: AVAILABLE }) {\n\t\t\t\tid\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/{____todos(condition:_{_user:_{_name:_"deep"_}_})_{_____id____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Query.todos.condition",
    "TodosCondition.user",
    "TodosConditionUser.name",
    "String",
    "Todo.id"
   ],
   "operation": "{\n\t\t\ttodos(condition: { user: { name: \"deep\" } }) {\n\t\t\t\tid\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/{____todos_{_____...TodoFragment____}___}___fragment_TodoFragment_on_Todo_{____id____text____user_{_____..._on_User_{______id______name_____}____}____done___}___ - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Todo.id",
    "Todo.text",
    "Todo.user",
    "User.id",
    "User.name",
    "Todo.done"
   ],
   "operation": "{\n\t\t\ttodos {\n\t\t\t\t...TodoFragment\n\t\t\t}\n\t\t}\n\t\tfragment TodoFragment on Todo {\n\t\t\tid\n\t\t\ttext\n\t\t\tuser {\n\t\t\t\t... on User {\n\t\t\t\t\tid\n\t\t\t\t\tname\n\t\t\t\t}\n\t\t\t}\n\t\t\tdone\n\t\t}\n\t\t"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/query_Todos($searchText:_String)_{____todos(condition:_{_searchText:_$searchText_})_{_____id____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Query.todos.condition",
    "TodosCondition.searchText",
    "String",
    "Todo.id"
   ],
   "operation": "query Todos($searchText: String) {\n\t\t\ttodos(condition: { searchText: $searchText }) {\n\t\t\t\tid\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestCreatedReports/query_Todos($userStatus:_TodosConditionUserStatus)_{____todos(condition:_{_userStatus:_$userStatus_})_{_____id____}___} - 1]
{
 "map": {
  "id": {
   "fields": [
    "Query.todos",
    "Query.todos.condition",
    "TodosCondition.userStatus",
    "TodosConditionUserStatus.AVAILABLE",
    "TodosConditionUserStatus.UNAVAILABLE",
    "Todo.id"
   ],
   "operation": "query Todos($userStatus: TodosConditionUserStatus) {\n\t\t\ttodos(condition: { userStatus: $userStatus }) {\n\t\t\t\tid\n\t\t\t}\n\t\t}"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "id",
   "timestamp": -1
  }
 ],
 "size": 1
}
---

[TestSendingQueuedReports - 1]
{
 "map": {
  "{ todos { id } } #1": {
   "fields": [
    "Query.todos",
    "Todo.id"
   ],
   "operation": "{ todos { id } } #1"
  },
  "{ todos { id } } #2": {
   "fields": [
    "Query.todos",
    "Todo.id"
   ],
   "operation": "{ todos { id } } #2"
  },
  "{ todos { id } } #3": {
   "fields": [
    "Query.todos",
    "Todo.id"
   ],
   "operation": "{ todos { id } } #3"
  },
  "{ todos { id } } #4": {
   "fields": [
    "Query.todos",
    "Todo.id"
   ],
   "operation": "{ todos { id } } #4"
  }
 },
 "operations": [
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "{ todos { id } } #1",
   "timestamp": -1
  },
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "{ todos { id } } #2",
   "timestamp": -1
  },
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "{ todos { id } } #3",
   "timestamp": -1
  },
  {
   "execution": {
    "duration": -1,
    "errorsTotal": 0,
    "ok": true
   },
   "metadata": {
    "client": {
     "name": "go-gqlhive",
     "version": "1.0.0"
    }
   },
   "operationMapKey": "{ todos { id } } #4",
   "timestamp": -1
  }
 ],
 "size": 4
}
---
